utils/sqlcommenter/src/SqlCommenter.hs

```diff
@@ -0,0 +1,268 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+-- See https://google.github.io/sqlcommenter/spec/
+-- for the specification of the SQL Commenter protocol.
+module SqlCommenter
+  ( sqlCommenter
+  , parseFirstSqlComment
+  -- * OpenTelemetry integration
+  , lookupSqlCommenterAttributes
+  , getSqlCommenterAttributes
+  , addTraceDataToAttributes
+  , getSqlCommenterAttributesWithTraceData
+  -- * Testing support
+  , urlEncode
+  ) where
+
+import Control.Applicative
+import Data.Attoparsec.Text
+import Data.Bits
+import qualified Data.ByteString as BS
+import Data.Char
+import Data.Foldable
+import Data.Function ((&))
+import Data.Text (Text)
+import Data.Map.Strict (Map)
+import qualified Data.Map.Strict as M
+import Data.Maybe
+import Data.Monoid
+import Data.Word
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
+import qualified Data.Text.IO as TIO
+import Prelude hiding (take, takeWhile)
+import Data.Text.Encoding (StrictBuilder)
+import qualified Data.Text.Internal.StrictBuilder as B
+import Network.HTTP.Types (urlDecode)
+import qualified OpenTelemetry.Context as Ctxt
+import qualified OpenTelemetry.Context.ThreadLocal as TL
+import OpenTelemetry.Propagator.W3CTraceContext (encodeSpanContext)
+import OpenTelemetry.Trace.Core as Core
+import System.IO.Unsafe
+
+sqlCommenterKey :: Ctxt.Key (Map Text Text)
+sqlCommenterKey = unsafePerformIO $ Ctxt.newKey "sqlcommenter-attributes"
+{-# NOINLINE sqlCommenterKey #-}
+
+lookupSqlCommenterAttributes :: Ctxt.Context -> Map Text Text
+lookupSqlCommenterAttributes ctxt = case Ctxt.lookup sqlCommenterKey ctxt of
+  Nothing -> mempty
+  Just attrs -> attrs
+
+getSqlCommenterAttributes :: IO (Map Text Text)
+getSqlCommenterAttributes = lookupSqlCommenterAttributes <$> TL.getContext
```

Comment:
I guess `sqlCommenterKey` has to be a global variable b/c of how `Vault` works, and since it's an implementation detail that never leaks beyond the scope of this module there shouldn't be a problem here?

checking my understanding: we can't make `sqlCommenterKey :: IO _` because we need `lookupSqlCommenterAttributes` to have a pure API for any callers who might need to use it in a pure context (even though `getSqlCommenterAttributes :: IO _`).

Reply:
Right. The main thing is that `sqlCommenterKey` needs to be _stable_ across uses. If we had users intitialize the key and pass it in any time they want to use it to look up the sqlcommenter attributes, then we could make it be an IO value. However, I don't think that's a great experience, and is relatively error prone compared to just making sure that there's a stable reference available.

Reply:
I didn't see this sort of context in `newKey` docs, so it might be worth adding to a Haddock on `sqlCommenterKey` at least. up to you, though